src/components/Conference.tsx

```diff
@@ -0,0 +1,211 @@
+import * as React from 'react';
+
+import {
+    ConferenceConnection,
+    IConfIncomingMessage,
+    IConfMessageSelf,
+    IConfIncomingMessageCandidate,
+    IConfIncomingMessageOffer,
+    IConfIncomingMessageAnswer,
+    IConfMessageAddPeer,
+    IConfMessageRemovePeer,
+    IConfOutgoingMessage
+} from '../data';
+import {
+    createOutgoingMessageJoin,
+    createOutgoingMessageCandidate,
+    createOutgoingMessageOffer,
+    createOutgoingMessageAnswer,
+    createOutgoingMessageBye
+} from '../services/ConferenceService';
+
+export interface IConferenceProps {
+    connect: () => ConferenceConnection;
+    room: string;
+}
+
+const userMediaConfig = {
+    audio: true,
+    video: true,
+}
+
+export class Conference extends React.Component<IConferenceProps, void> {
+    private connection: ConferenceConnection;
+    private peerConnections: { [id: string]: RTCPeerConnection };
+    private localStream: MediaStream;
+    private remoteStreams: { [id: string]: MediaStream };
+    private localId: string;
+
+    constructor() {
+        super();
+        this.connection = this.props.connect();
+        this.connection.subscribe(this.handleIncomingMessage);
+        this.handleIncomingMessage.bind(this);
+        this.joinRoom(this.props.room);
+        this.getUserMedia();
```

Comment:
There is a race condition here. You should only call `this.connection.subscribe` after getUserMedia has set the local stream